File: Data Structures & Algorithms/buy-and-sell-crypto-with-cooldown/submission-0.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        n = len(prices)
        dp = [-1] * n
        def dfs(i, buy, summ):
            if i == n-1:
                return summ+prices[i]-buy if buy != -1 and prices[i] > buy else summ
            elif i >= n:
                return summ
                
            ret = -1
             # Skip the current day altogether
            print(f'i: {i}, buy: {buy}, summ: {summ}')
            ret = max(ret, dfs(i+1, buy, summ))

            if buy == -1: # Buy current day, then skip a day
                print(f'i: {i}, buy: {buy}, summ: {summ}')
                ret = max(ret, dfs(i+1, prices[i], summ))

            elif prices[i] > buy: # Sell coin
                print(f'i: {i}, buy: {buy}, summ: {summ}')
                ret = max(ret, dfs(i+2, -1, summ+prices[i]-buy))

            return ret
            
        return dfs(0, -1, 0)

File: Data Structures & Algorithms/pow-x-n/submission-0.cpp
class Solution {
public:
    double myPow(double x, int n) {
        if (n == 0) return 1.0;
        if (n == 1) return x;
        if (n < 0) return myPow(1/x, -n);
        if (n % 2 == 0) return myPow(x*x, n/2);
        return myPow(x*x, n/2) * x;
    }
};
